streamer/src/nonblocking/stream_throttle.rs

```diff
@@ -208,19 +209,24 @@ impl ConnectionStreamCounter {
         }
     }
 
-    pub(crate) fn reset_throttling_params_if_needed(&self) {
-        const THROTTLING_INTERVAL: Duration = Duration::from_millis(STREAM_THROTTLING_INTERVAL_MS);
-        if tokio::time::Instant::now().duration_since(*self.last_throttling_instant.read().unwrap())
-            > THROTTLING_INTERVAL
+    /// Reset the counter and last throttling instant and
+    /// return last_throttling_instant regardless it is reset or not.
+    pub(crate) fn reset_throttling_params_if_needed(&self) -> tokio::time::Instant {
+        let last_throttling_instant = *self.last_throttling_instant.read().unwrap();
+        if tokio::time::Instant::now().duration_since(last_throttling_instant)
+            > STREAM_THROTTLING_INTERVAL
         {
             let mut last_throttling_instant = self.last_throttling_instant.write().unwrap();
```

Comment:
It was already the case before this PR so doesn't block this PR, but this is
racy: we're doing a check with a read lock, then dropping and acquiring a write
lock. Multiple threads can find that they're > THROTTLING_INTERVAL with the read
lock and update the last throttling instant.

I think that this connection-wide limit can be enforced without locks. Every connection task keeps
its last_throttling_instant and stream_count. The max streams allowed per connection is scaled by
the number of connections opened by a peer, which can be passed as an atomic to each connection
task.

Reply:
I created a separate issue to address as it is orthogonal to this PR.

Reply:
https://github.com/anza-xyz/agave/issues/942

Reply:
It should not be racy, as we check again for the timeout after acquiring the write lock. Only one thread will acquire the write lock, and it'll update the `last_throttling_instant`.